File: ImgConcat.Tests/FcpxmlBuilderTests.cs
using System.Xml.Linq;
using NUnit.Framework;

namespace ImgConcat.Tests
{
    public class FcpxmlBuilderTests
    {
        private static FcpxmlOptions DefaultOptions(int crossfadeFrames = 15) =>
            new("Test Project", 1920, 1080, 30, 60, crossfadeFrames);

        private static XDocument BuildDocument(IReadOnlyList<FcpxmlStill> stills, FcpxmlOptions options) =>
            XDocument.Parse(FcpxmlBuilder.Build(stills, options));

        private static List<FcpxmlStill> Stills(int count) =>
            Enumerable.Range(1, count)
                .Select(i => new FcpxmlStill(Path.Combine(Path.GetTempPath(), $"frame {i}.jpg"), 4000, 3000))
                .ToList();

        [Test]
        public void Build_ShouldPlaceEachStillOnSpineBackToBack()
        {
            var doc = BuildDocument(Stills(3), DefaultOptions());

            var clips = doc.Descendants("spine").Elements("video").ToList();
            Assert.That(clips, Has.Count.EqualTo(3));
            Assert.That(clips.Select(c => c.Attribute("offset")!.Value), Is.EqualTo(new[] { "0s", "60/30s", "120/30s" }));
            Assert.That(clips.All(c => c.Attribute("duration")!.Value == "60/30s"), Is.True);

            var sequence = doc.Descendants("sequence").Single();
            Assert.That(sequence.Attribute("duration")!.Value, Is.EqualTo("180/30s"));
        }

        [Test]
        public void Build_ShouldCentreCrossDissolveOnEachCut()
        {
            var doc = BuildDocument(Stills(3), DefaultOptions(crossfadeFrames: 15));

            var transitions = doc.Descendants("spine").Elements("transition").ToList();
            Assert.That(transitions, Has.Count.EqualTo(2));
            Assert.That(transitions.Select(t => t.Attribute("offset")!.Value), Is.EqualTo(new[] { "53/30s", "113/30s" }));
            Assert.That(transitions.All(t => t.Attribute("duration")!.Value == "15/30s"), Is.True);

            var effect = doc.Descendants("effect").Single();
            Assert.That(effect.Attribute("uid")!.Value, Is.EqualTo(FcpxmlBuilder.CrossDissolveUid));
            Assert.That(transitions.All(t => t.Element("filter-video")!.Attribute("ref")!.Value == effect.Attribute("id")!.Value), Is.True);
        }

        [Test]
        public void Build_WithoutCrossfade_ShouldOmitTransitions()
        {
            var doc = BuildDocument(Stills(3), DefaultOptions(crossfadeFrames: 0));

            Assert.That(doc.Descendants("transition"), Is.Empty);
            Assert.That(doc.Descendants("effect"), Is.Empty);
        }

        [Test]
        public void Build_ShouldReferenceAssetsWithFileUrlsAndUniqueIds()
        {
            var stills = Stills(2);
            var doc = BuildDocument(stills, DefaultOptions());

            var assets = doc.Descendants("asset").ToList();
            Assert.That(assets, Has.Count.EqualTo(2));
            Assert.That(assets.Select(a => a.Element("media-rep")!.Attribute("src")!.Value),
                Is.EqualTo(stills.Select(s => new Uri(s.Path).AbsoluteUri)));
            Assert.That(assets[0].Element("media-rep")!.Attribute("src")!.Value, Does.StartWith("file://").And.Contain("frame%201.jpg"));

            var ids = doc.Descendants().Attributes("id").Select(a => a.Value).ToList();
            Assert.That(ids, Is.Unique);

            var clipRefs = doc.Descendants("spine").Elements("video").Select(v => v.Attribute("ref")!.Value);
            Assert.That(clipRefs, Is.EqualTo(assets.Select(a => a.Attribute("id")!.Value)));
        }

        [Test]
        public void Build_ShouldShareStillFormatsBySize()
        {
            var stills = new List<FcpxmlStill>
            {
                new("/tmp/a.jpg", 4000, 3000),
                new("/tmp/b.jpg", 4000, 3000),
                new("/tmp/c.jpg", 3000, 4000),
            };
            var doc = BuildDocument(stills, DefaultOptions());

            var stillFormats = doc.Descendants("format").Where(f => f.Attribute("name")!.Value == "FFVideoFormatRateUndefined").ToList();
            Assert.That(stillFormats, Has.Count.EqualTo(2));

            var sequenceFormat = doc.Descendants("format").First();
            Assert.That(sequenceFormat.Attribute("name")!.Value, Is.EqualTo("FFVideoFormat1080p30"));
            Assert.That(sequenceFormat.Attribute("frameDuration")!.Value, Is.EqualTo("1/30s"));
        }

        [Test]
        public void Build_ShouldProduceVersionedDocumentWithDoctype()
        {
            var xml = FcpxmlBuilder.Build(Stills(1), DefaultOptions());

            Assert.That(xml, Does.Contain("<!DOCTYPE fcpxml>"));
            var doc = XDocument.Parse(xml);
            Assert.That(doc.Root!.Name.LocalName, Is.EqualTo("fcpxml"));
            Assert.That(doc.Root.Attribute("version")!.Value, Is.EqualTo(FcpxmlBuilder.Version));
            Assert.That(doc.Descendants("project").Single().Attribute("name")!.Value, Is.EqualTo("Test Project"));
        }

        [Test]
        public void Build_WithNoStills_ShouldThrow()
        {
            Assert.Throws<ArgumentException>(() => FcpxmlBuilder.Build(new List<FcpxmlStill>(), DefaultOptions()));
        }

        [TestCase("video", OutputFormat.Video)]
        [TestCase("MP4", OutputFormat.Video)]
        [TestCase("fcpxml", OutputFormat.FinalCutPro)]
        [TestCase("fcp", OutputFormat.FinalCutPro)]
        [TestCase("both", OutputFormat.Both)]
        public void OutputFormatParser_ShouldParseKnownValues(string value, OutputFormat expected)
        {
            Assert.That(OutputFormatParser.TryParse(value, out var format), Is.True);
            Assert.That(format, Is.EqualTo(expected));
        }

        [Test]
        public void OutputFormatParser_ShouldRejectUnknownValues()
        {
            Assert.That(OutputFormatParser.TryParse("gif", out _), Is.False);
        }
    }
}

File: StopMotionApp/Tests/StopMotionKitTests/FCPXMLBuilderTests.swift
import XCTest
@testable import StopMotionKit

final class FCPXMLBuilderTests: XCTestCase {
    private func stills(_ count: Int) -> [FCPXMLStill] {
        (1...count).map { FCPXMLStill(url: URL(fileURLWithPath: "/tmp/frame \($0).jpg"), width: 4000, height: 3000) }
    }

    private func document(stills: [FCPXMLStill], crossfadeFrames: Int = 15) throws -> XMLDocument {
        let builder = FCPXMLBuilder(
            projectName: "Test & Project",
            width: 1920,
            height: 1080,
            timing: FrameTiming(frameRate: 30, framesPerImage: 60, crossfadeFrames: crossfadeFrames)
        )
        return try XMLDocument(xmlString: try builder.build(stills: stills))
    }

    private func values(_ document: XMLDocument, _ xpath: String) throws -> [String] {
        try document.nodes(forXPath: xpath).compactMap(\.stringValue)
    }

    func testClipsAreBackToBack() throws {
        let doc = try document(stills: stills(3))
        XCTAssertEqual(try values(doc, "//spine/video/@offset"), ["0s", "60/30s", "120/30s"])
        XCTAssertEqual(try values(doc, "//spine/video/@duration"), ["60/30s", "60/30s", "60/30s"])
        XCTAssertEqual(try values(doc, "//sequence/@duration"), ["180/30s"])
    }

    func testCrossDissolveIsCentredOnEachCut() throws {
        let doc = try document(stills: stills(3))
        XCTAssertEqual(try values(doc, "//spine/transition/@offset"), ["53/30s", "113/30s"])
        XCTAssertEqual(try values(doc, "//spine/transition/@duration"), ["15/30s", "15/30s"])
        XCTAssertEqual(try values(doc, "//effect/@uid"), [FCPXMLBuilder.crossDissolveUID])
        let effectID = try values(doc, "//effect/@id").first
        XCTAssertEqual(Set(try values(doc, "//transition/filter-video/@ref")), [effectID!])
    }

    func testNoCrossfadeMeansHardCuts() throws {
        let doc = try document(stills: stills(3), crossfadeFrames: 0)
        XCTAssertTrue(try doc.nodes(forXPath: "//transition").isEmpty)
        XCTAssertTrue(try doc.nodes(forXPath: "//effect").isEmpty)
    }

    func testAssetsUseFileURLsAndUniqueIDs() throws {
        let doc = try document(stills: stills(2))
        let sources = try values(doc, "//asset/media-rep/@src")
        XCTAssertEqual(sources, ["file:///tmp/frame%201.jpg", "file:///tmp/frame%202.jpg"])

        let ids = try values(doc, "//@id")
        XCTAssertEqual(ids.count, Set(ids).count)
        XCTAssertEqual(try values(doc, "//spine/video/@ref"), try values(doc, "//asset/@id"))
    }

    func testFormatsAndEscaping() throws {
        let doc = try document(stills: [
            FCPXMLStill(url: URL(fileURLWithPath: "/tmp/a.jpg"), width: 4000, height: 3000),
            FCPXMLStill(url: URL(fileURLWithPath: "/tmp/b.jpg"), width: 4000, height: 3000),
            FCPXMLStill(url: URL(fileURLWithPath: "/tmp/c.jpg"), width: 3000, height: 4000)
        ])
        XCTAssertEqual(try values(doc, "/fcpxml/@version"), [FCPXMLBuilder.version])
        XCTAssertEqual(try values(doc, "//format[@id='r1']/@name"), ["FFVideoFormat1080p30"])
        XCTAssertEqual(try values(doc, "//format[@id='r1']/@frameDuration"), ["1/30s"])
        XCTAssertEqual(try doc.nodes(forXPath: "//format[@name='FFVideoFormatRateUndefined']").count, 2)
        XCTAssertEqual(try values(doc, "//project/@name"), ["Test & Project"])
    }

    func testEmptyInputThrows() {
        let builder = FCPXMLBuilder(projectName: "Empty", width: 1920, height: 1080, timing: FrameTiming(frameRate: 30, framesPerImage: 30, crossfadeFrames: 0))
        XCTAssertThrowsError(try builder.build(stills: [])) { error in
            XCTAssertEqual(error as? FCPXMLBuilder.BuildError, .noStills)
        }
    }
}
